File: questions/1655. Distribute Repeating Integers/distri.py
class Solution:
    def canDistribute(self, nums: List[int], quantity: List[int]) -> bool:
        ma = defaultdict(int)
        for i in nums:
            ma[i] += 1
        count = list(Counter(ma).values())
        count.sort()
        quantity.sort(reverse = True)
        def dfs(c, q, l):
            if l == len(q):
                return True
            for i in range(len(c)):
                if c[i] >= q[l]:
                    c[i] -= q[l]
                    if dfs(c, q, l + 1):
                        return True
                    c[i] += q[l]
            return False
        return dfs(count, quantity, 0)
        

File: questions/1655. Distribute Repeating Integers/distri.cpp
class Solution {
public:
    bool canDistribute(vector<int>& nums, vector<int>& quantity) {
        unordered_map<int, int> m;
        for(int i: nums){
            m[i] += 1;
        }
        vector<int> count;
        for(auto i = m.begin(); i != m.end(); ++i){
            count.push_back(i->second);
        }
        sort(count.begin(), count.end());
        sort(quantity.begin(), quantity.end());
        reverse(quantity.begin(), quantity.end());
        return dfs(count, quantity, 0);
    }
    bool dfs(vector<int>& count, vector<int>& quantity, int level){
        if(level == quantity.size()){
            return true;
        }
        for(auto it = count.begin(); it != count.end(); ++it){
            if(*it >= quantity[level]){
                *it -= quantity[level];
                if(dfs(count, quantity, level + 1)){
                    return true;
                }
                *it += quantity[level];
            }
        }
        return false;
    }
};
